examples/hosts/app-integration/document-recovery/src/recoveryAgent.ts

```diff
@@ -0,0 +1,145 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { EventEmitter } from "events";
+import { ContainerSchema, SharedMap } from "fluid-framework";
+import {
+    AzureClient,
+    AzureContainerVersion,
+} from "@fluidframework/azure-client";
+import { connectionConfig } from "./azureConfig";
+import { counterValueKey } from "./dataController";
+
+export type RecoveryStatus =
+    | "NotStarted"
+    | "KickedOff"
+    | "Success"
+    | "Error"
+    | "Cancelled";
+
+export interface RecoveryInfo {
+    originalContainerId: string;
+    isContainerRecovered?: boolean;
+    recoveredContainerId?: string;
+    recoveryStatus?: RecoveryStatus;
+    recoveryLog: string;
+}
+
+export class RecoveryAgent extends EventEmitter {
+    private readonly azureClient: AzureClient;
+    private readonly maxDocVersions = 5;
+    private recoveryLog: string;
+    private recoveryStatus: RecoveryStatus;
+    public recoveredContainerId: string | undefined;
+
+    constructor(
+        private readonly orgContainerId: string,
+        private readonly orgContainerSchema: ContainerSchema,
+    ) {
+        super();
+        const clientProps = {
+            connection: connectionConfig,
+        };
+        this.azureClient = new AzureClient(clientProps);
+        this.recoveredContainerId = undefined;
+        this.recoveryLog = "";
+        this.recoveryStatus = "NotStarted";
+    }
+
+    public static createRecoveryAgent(
+        containerId: string,
+        containerSchema: ContainerSchema,
+    ): RecoveryAgent {
+        return new RecoveryAgent(containerId, containerSchema);
+    }
+
+    public get getRecoveryInfo(): RecoveryInfo {
+        if (!this.orgContainerId) {
+            throw new Error("Cannot retreive status: Missing container ID.");
+        }
+
+        return {
+            originalContainerId: this.orgContainerId,
+            recoveryStatus: this.recoveryStatus,
+            recoveryLog: this.recoveryLog,
+            recoveredContainerId: this.recoveredContainerId,
+        };
+    }
+
+    public async startRecovery(): Promise<void> {
+        Promise.race([
+            this.recoverDoc(),
+            new Promise((_resolve, reject) =>
+                setTimeout(() => reject(new Error("timeout")), 1000),
+            ),
+        ]).then((val) => {
```

Comment:
might be good to store the timeout handle in a variable and add `clearTimeout(timeoutHandle)` here so that if the `recoverDoc()` promise finishes first the timer doesn't fire redundantly. Just being pedantic :)